kratos/containers/container_expression/expressions/expression_iterator.h

```diff
@@ -0,0 +1,283 @@
+        result.mEntityDataBeginIndex = result.mEntityIndex * result.mFlattenedShapeSize;
+        return result;
+    }
+
+    ///@}
+    ///@name Public operators
+    ///@{
+
+    double operator*() const
+    {
+        return mpExpression->Evaluate(mEntityIndex, mEntityDataBeginIndex, mComponentIndex);
+    }
+
+    bool operator==(const ConstIteratorType& rOther) const
+    {
+        return (&*mpExpression == &*rOther.mpExpression) && (mEntityIndex == rOther.mEntityIndex) && (mComponentIndex == rOther.mComponentIndex);
+    }
+
+    bool operator!=(const ConstIteratorType& rOther) const
+    {
+        return !this->operator==(rOther);
+    }
+
+    ConstIteratorType operator+(const IndexType n) const
+    {
+        ConstIteratorType result(mpExpression);
+        result.mComponentIndex = (mComponentIndex + n) % mFlattenedShapeSize;
+        result.mEntityIndex += n / mFlattenedShapeSize;
+        result.mEntityDataBeginIndex = result.mEntityIndex * mFlattenedShapeSize;
+        return result;
+    }
+
+    ConstIteratorType& operator++()
+    {
+        ++mComponentIndex;
+        if (mComponentIndex == mFlattenedShapeSize) {
+            mComponentIndex = 0;
+            ++mEntityIndex;
+            mEntityDataBeginIndex = mEntityIndex * mFlattenedShapeSize;
+        }
+        return *this;
+    }
+
+    ConstIteratorType operator++(int)
+    {
+        ConstIteratorType temp = *this;
+        ++*this;
+        return temp;
+    }
+
+    ///@}
+
+private:
+    ///@name Private member variables
+    ///@{
+
+    Expression::Pointer mpExpression;
+
+    IndexType mEntityIndex;
+
+    IndexType mEntityDataBeginIndex;
+
+    IndexType mComponentIndex;
+
+    IndexType mFlattenedShapeSize;
+
+    ///@}
+};
+
+/**
+ * @brief Iterator class for expressions with concrete data arrays
+ *
+ */
+template<>
+class ExpressionIterator<LiteralFlatExpression<char>> {
+public:
+    ///@name Type definitions
+    ///@{
+
+    using ExpressionType = LiteralFlatExpression<char>;
+
+    using DataType = typename ExpressionType::DataType;
+
+    using IteratorType = DataType*;
+
+    using ConstIteratorType = DataType const*;
+
+    KRATOS_CLASS_POINTER_DEFINITION(ExpressionIterator);
+
+    ///@}
+    ///@name Life cycle
+    ///@{
+
+    ExpressionIterator(typename ExpressionType::Pointer pExpression): mpExpression(pExpression) {}
+
+    ///@}
+    ///@name Public operations
+    ///@{
+
+    typename ExpressionType::Pointer GetExpression() { return mpExpression; }
+
+    IteratorType begin() { return mpExpression->begin(); }
+
+    IteratorType end() { return mpExpression->end(); }
+
+    ConstIteratorType cbegin() { return mpExpression->cbegin(); }
+
+    ConstIteratorType cend() { return mpExpression->cend(); }
+
+    ///@}
+
+private:
+    ///@name Private member variables
+    ///@{
+
+    typename ExpressionType::Pointer mpExpression;
+
+    ///@}
+};
+
+template<>
+class ExpressionIterator<LiteralFlatExpression<int>> {
```

Comment:
I think macros are OK to be used here to eliminate code duplication.

Reply:
removed since I moved the iterators to the respective classes.